V6 (Chakra UI V3)

csandman opened the pull request:
~~This is still very much a work in progress, I'm just going to leave this open in draft state while I work on it.~~

This version is now in the final stages, and should be fully production ready! For anyone interested in upgrading to Chakra v3, you can install the latest version of this package by using the `rc` tag:

```sh
npm i chakra-react-select@rc
# or
yarn add chakra-react-select@rc
```

I included a checklist of remaining items that need to be finished before publishing a full release at the bottom of this post. At this point, it's just documentation changes.

Here is a list of breaking changes that should be kept in mind when upgrading:

- All of the styles pulled from the end user's theme are now using styles from the [`select` component](https://www.chakra-ui.com/docs/components/select) (except for the control which still uses input styles).
- Everything that was `colorScheme` is now `colorPalette`
  - `tagColorScheme` -> `tagColorPalette`
  - `selectedOptionColorScheme` -> `selectedOptionColorPalette`
  - An individual option's `colorScheme` -> `colorPalette`
- Some props were replaced on `chakraComponents.LoadingIndicator`:
  - `emptyColor` -> `trackColor` (this isn't a real prop, but it maps to the `--spinner-track-color` css variable).
  - `speed` -> `animationDuration`
  - `thickness` -> `borderWidth`
  - The prop `colorPalette` was also added to the loading indicator which can be used to set the color without specifying a token number (responds better to changes between light and dark mode).
- `focusBorderColor` was replaced with `focusRingColor` and `errorBorderColor` was removed, as there is no longer any equivalent prop that exists.
- All boolean props have had the `is` prefix removed:
  - `isRequired` is now the `required` attribute that's already built-in to `react-select`.
  - `isReadOnly` -> `readOnly`
  - `isInvalid` -> `invalid`
  - Adds `disabled`, which will override the `isDisabled` prop built-in to `react-select`
    - This is just to keep the props consistent with each other.
- Add a default for the `menuPlacement` of `"auto"`. This is to more closely match the default flipping behavior of the Chakra `Select`. 

---

## Remaining Tasks
- [x] Update the docs for all of the custom props
- [x] Update the demo to include examples for all of the custom props
- [x] Replace all images in the readme
- [x] Fix the `useSize` hook
- [x] Update stack blitz demo links in the readme
- [x] Check all readme copy to ensure everything is updated to v3

csandman committed:
Initial commit of v6

csandman committed:
6.0.0-next.0

csandman committed:
Replace chakra.span with Span

csandman committed:
Fix subtle styling

csandman committed:
Make the default size getter not fail without a value

csandman committed:
Remove unused eslint rule

csandman committed:
Merge remote-tracking branch 'origin' into v6

csandman committed:
Add `next` tag to the publish config

csandman committed:
6.0.0-next.1

csandman committed:
Merge remote-tracking branch 'origin' into v6

csandman committed:
Upgrade to real Chakra v3

csandman committed:
Bump up to official Chakra v3

csandman committed:
Update dependencies

csandman committed:
Clean up size prop again

csandman committed:
A little more cleanup

csandman committed:
Add an invalid example

csandman committed:
Fix placeholder color

csandman committed:
Remove default cursors

csandman committed:
6.0.0-next.2

csandman committed:
Move cursors to correct location

csandman committed:
Remove extra multi-value styles

csandman committed:
Add other size examples

csandman committed:
Fix clear button height

csandman committed:
Fix clear indicator sizing

csandman committed:
Remove duplicate aria-label

csandman committed:
Add note about clear indicator

csandman committed:
Remove redundant duplicate styles from the control

csandman committed:
Make the clear indicator icon slightly larger

csandman committed:
Merge branch 'main' into v6

csandman committed:
Update prop typing comments, and reorganize useChakraSelectProps

csandman committed:
Update demo

csandman committed:
Remove theme color palette

mskasal commented:
@csandman  Hello there. Do we have a timeline for migrating to ChakraUI v3? And how can we help?

csandman committed:
Fix tags postpublish script

csandman committed:
Clean up ignore files

csandman committed:
Remove extra variant

csandman committed:
Update demo a little

csandman committed:
Remove `react-dom` as a peer dependency

csandman committed:
6.0.0-next.3

csandman commented:
> @csandman Hello there. Do we have a timeline for migrating to ChakraUI v3? And how can we help?

Hey, sorry for the delay in communication, I've been very busy recently and haven't had as much time to focus on this project as I'd like. Let me give a rundown of the current situation.

So, for full transparency, my company isn't planning to upgrade to Chakra V3 in the near future due to the dramatic size of the refactor necessary. That means that I will most likely not be personally using this new version of `chakra-react-select` in production for the forseeable future. Because of that, it's made me pretty hesitant to mark this as an official release, as I haven't really had a chance to stress test it as much as I'd like.

Besides that, the biggest blocker really is updating the docs. With all of the changes to the core component APIs along with the theming API in Chakra v3, I will have to do a full refresh of the docs for this project to reflect those changes, and how this package utilizes them. Overall, it's not actually too complicated past the changes I mentioned above, I just wrote a giant wall of text for the docs over time, and it's going to take a while to comb through them haha.

---

Now, all that being said, I think the changes in the [`6.0.0-next.3`](https://www.npmjs.com/package/chakra-react-select/v/6.0.0-next.3) are pretty much production ready! The best thing people can do to help get this fully released is to give it a try in a real app and give some feedback on it. That feedback can be anything from "it's working fine" to "it's totally broken in this use case", to "it's missing this core feature that would be great for v3 compatibility". If it does work fine, that feedback will help give me the confidence to put an official release tag on it and ship it. It's just been tough for me to call it an official release because I haven't really heard any feedback on this pre-release so far, positive or negative.

And if there is any feedback around things genuinely not working properly, I will do my best to get things fixed in a timely manner to avoid slowing anyone's development down. If you post an issue on a weekday, I should have a fix ready by that night, if not the next night.

b4stien commented:
The `next-themes` peer dependency is a bit too restrictive, chakra (3) snippet CLI asks for `next-themes` `^0.4.3`. Maybe you can align with them?

csandman commented:
@b4stien Good call, I intended for that version `^0.3.0` to match all minor versions above `0.3.0`, but I forgot that the `^` flag doesn't count the second version num as a minor version when the major version is 0. I'll deploy a new version with more lenient version matching in a little bit.

csandman committed:
Start working on updating the docs

csandman committed:
Generalize `next-themes` peer dep version

csandman committed:
6.0.0-next.4

csandman commented:
@b4stien Ok I just pushed a new version that changes the peer dependency requirement of `next-themes` to `0.x`, so hopefully that fixes you problem! The change is live in `v6.0.0-next.4`

sblyvwx commented:
@csandman Hi, I just wanted to mention that I encountered runtime errors on my side if any of the Select components weren't enclosed within the Chakra's Field component. The error was "TypeError: Cannot read properties of undefined (reading 'readOnly')", meaning that inputProps.readOnly and similar properties were undefined.

csandman commented:
> @csandman Hi, I just wanted to mention that I encountered runtime errors on my side if any of the Select components weren't enclosed within the Chakra's Field component. The error was "TypeError: Cannot read properties of undefined (reading 'readOnly')", meaning that inputProps.readOnly and similar properties were undefined.

Ok thanks for letting me know, I'll take a look tonight! Can you provide an example of your implementation of it that I can go off of?

csandman committed:
Add proper handling for no `Field.Root` wrapper

csandman committed:
6.0.0-next.5

csandman committed:
Fix the typing on the `useFieldContext` return

dontic commented:
> That feedback can be anything from "it's working fine"

I just implemented a pretty complex case with debouncing using `6.0.0-next.5` like this and it's working like a charm:

```typescript
// Async Listing select stuff
const [listingValue, setListingValue] = useState<any>(null);

const _loadListingOptions = (inputValue: string, callback: any) => {
  autovisitaListingsList({ search: inputValue }).then((response) => {
    const newOptions = response.map((listing) => ({
      label: listing.title,
      value: listing.pk
    }));
    callback(newOptions);
  });
};

const loadListingOptions = useDebouncedCallback(_loadListingOptions, 500);

{/* ... */}

{/* Listing */}
<Field
  label="Listing"
  invalid={errors.listing ? true : false}
  errorText={errors.listing?.message}
>
  <AsyncSelect
    name="listing"
    placeholder="Select listing"
    loadOptions={loadListingOptions}
    noOptionsMessage={() => "Start typing to search"}
    loadingMessage={() => "Loading..."}
    onChange={(value: any) => {
      setListingValue(value); // Set value to local state
      setValue("listing", value?.value); // Send value to hook form
    }}
    value={listingValue}
  />

  <Button
    size={"xs"}
    variant={"subtle"}
    colorPalette="gray"
    onClick={() => {
      setValue("listing", null);
      setListingValue(null);
    }}
  >
    Limpiar
  </Button>
</Field>
```

> Note: I'm controlling `listingValue` so I can pass it initial data gathered from my API.

csandman commented:
@dontic @lovegehlin Thanks for the feedback! I'm working on the changes to the docs now, so I should have the first full release ready pretty soon!

csandman committed:
Update demo to match previous demo

csandman committed:
Re-add color mode from snippet

csandman committed:
Use theme styles for item group

csandman committed:
Add missing demos

csandman committed:
Add focusRingColor prop and re-add custom spinner props

csandman committed:
Start on updating readme

csandman committed:
Properly wrap tag end element

csandman committed:
Move end element wrapper inside tag remove

csandman committed:
Add extra line to the readme

csandman committed:
Fix placeholder color

csandman committed:
Fix font sizing/color for empty messages

csandman committed:
Update tag to rc

csandman committed:
6.0.0-rc.0

csandman commented:
Good news, I spend a bunch of time tonight going through the docs and updating a large portion of it. I updated the docs for all of the custom props provided by this package, as well as the correlations between the `react-select` components and the Chakra theme styles they are now pulling from. In the process, I also updated the demo to include examples for all of the custom props, giving me the opportunity to test everything more thoroughly. I caught a few more issues with the styling that are now also resolved, and the demo should look more "real" now for anyone who wants to test it.

If you do want to test it, you can use the [`chakra-react-select-demo`](https://pkg.pr.new/template/9ba69fad-22c9-400d-aeaa-bc9171d4326f) link from the @pkg-pr-new comment above: https://github.com/csandman/chakra-react-select/pull/341#issuecomment-2430618608

After these changes, I'm much more confident in the current state of this branch, so I also released the first release candidate version, `v6.0.0-rc.0`! There aren't too many obvious differences, but I did fix a few styles, and I added the prop `focusRingColor` to replace the previously removed `focusBorderColor`. There was also a font-sizing issue fixed with the empty messages, as well as a new wrapping "end element" for the tag remove button.

I still have some left to do to get the readme docs up to date, including replacing all of the examples, but it should be mostly possible to figure everything out based on the other changes to the docs, which you can see here: https://github.com/csandman/chakra-react-select/tree/v6?tab=readme-ov-file#chakra-react-select-v6

And besides that, I plan to put together a more comprehensive change list to be provided with the release notes.

csandman committed:
Add custom spinner example

csandman committed:
Update readme for spinner props

csandman committed:
Update MultiValue custom components to use real tag components

csandman committed:
Add multiple advanced examples

csandman committed:
Remove CodeSandbox examples

csandman committed:
A bit more readme cleanup

csandman committed:
A bit more readme cleanup

csandman committed:
Fix popover example

csandman committed:
Go back to the original non-chakra components for full styling control

csandman committed:
Expand react dependency range to include 19

csandman committed:
6.0.0-rc.1

csandman committed:
Fix the multi value tag styles once more

csandman committed:
6.0.0-rc.2

csandman committed:
Clean up dropdown indicator + icon styles

csandman committed:
Remove extra styles from the root container

csandman committed:
Fix connected menu example

csandman committed:
Re-add style to remove gap

csandman committed:
Remove more redundant styles

csandman committed:
Re-add position relative on root container

csandman committed:
Move padding styles back to their original position

JAD3N commented:
The `react-select` dependency on the V6 branch has a peer dependency of up-to React 18. Would it be fine to bump the version to 5.9.0 so there are no peer dependency warnings when using React 19?

csandman committed:
Fix eslint config for demo

csandman committed:
Add workspace setting for TypeScript version

csandman committed:
Update all dependency versions

csandman committed:
6.0.0-rc.3

csandman commented:
> The `react-select` dependency on the V6 branch has a peer dependency of up-to React 18. Would it be fine to bump the version to 5.9.0 so there are no peer dependency warnings when using React 19?

@JAD3N Good callout, I had seen the release notification but forgot to handle the implementation. I just went ahead and published `v6.0.0-rc.3` which updated the peer dependency for `react-select` to be a bit more lenient, and allow for `react-select ^5.8.0` (anything in the 5.x range above 5.8).

In order to ensure everything still works, I also went ahead and updated all of the dependencies of the demo so that is now also on `react@19`, and everything is working like a charm! Give it a try when you get a chance.

csandman committed:
Update react-select dependency to have `5.9.0` as the minimum version

csandman committed:
6.0.0-rc.4

csandman committed:
update demo package-lock.json

csandman committed:
Add placeholder defaultSize demo for select

csandman committed:
Fix the `useSize` hook

csandman committed:
Remove inferred sizing for the spinner

csandman committed:
Simplify the message padding values

csandman committed:
Remove other example links and use built-in Chakra Z indices for portal examples

csandman committed:
Remove advanced examples section from the Readme

csandman committed:
Update readme to be less first person

csandman committed:
Remove unused `@types/react-dom`

csandman committed:
Update react version in demo eslint config

csandman committed:
Update eslint config to the latest

csandman committed:
Run a fresh pacakge-lock install

csandman commented:
> I have been using rc.4 and it has been working great so far. I had to adjust some of the old styling, but the it was more of removing them since the default out of the box in chakra v3 works great. 

Thanks for the feedback! When you say "removing them", do you mean removing custom styling you previously had using the chakraStyles prop or something?

csandman committed:
Update note formatting

csandman committed:
Finish updating the README

csandman commented:
Ok all, the official release is most likely coming Monday! Thanks for all the support!

maulik13 commented:
> > I have been using rc.4 and it has been working great so far. I had to adjust some of the old styling, but the it was more of removing them since the default out of the box in chakra v3 works great.
> 
> Thanks for the feedback! When you say "removing them", do you mean removing custom styling you previously had using the chakraStyles prop or something?
> 

Sorry for the late reply, I meant that the old styling was specific to chakra-v2 customization to match my app. It did not break anything (if I remember correctly), but it was more of just a small clean up. Looking forward to the new release :D

csandman committed:
Add ability to define custom variants

csandman committed:
6.0.0-rc.5

csandman committed:
Update demo package lock

csandman committed:
Remove rc tag from package.json

csandman committed:
Update all stackblitz links

csandman commented:
v6 is officially released, give it a try and let me know if anything isn't working right!
